src/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD113CheckForSystemIAsyncDisposableAnalyzerTests.cs

```diff
@@ -0,0 +1,271 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+
+namespace Microsoft.VisualStudio.Threading.Analyzers.Tests
+{
+    using System.Threading.Tasks;
+    using Xunit;
+    using VBVerify = VisualBasicCodeFixVerifier<VSTHRD113CheckForSystemIAsyncDisposableAnalyzer, Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+    using Verify = CSharpCodeFixVerifier<VSTHRD113CheckForSystemIAsyncDisposableAnalyzer, Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+    public class VSTHRD113CheckForSystemIAsyncDisposableAnalyzerTests
+    {
+        private const string Preamble = @"
+using System.Threading.Tasks;
+using BclAsyncDisposable = System.IAsyncDisposable;
+using VsThreadingAsyncDisposable = Microsoft.VisualStudio.Threading.IAsyncDisposable;
+";
+
+        private const string VBPreamble = @"
+Imports System.Threading.Tasks
+Imports BclAsyncDisposable = System.IAsyncDisposable
+Imports VsThreadingAsyncDisposable = Microsoft.VisualStudio.Threading.IAsyncDisposable
+";
+
+        [Fact]
+        public async Task MethodChecksBoth_WithIsCast()
+        {
+            var test = Preamble + @"
+class Test {
+    async Task CheckAndDispose(object o) {
+        if (o is BclAsyncDisposable bcl) {
+            await bcl.DisposeAsync();
+        }
+        else if (o is VsThreadingAsyncDisposable vs) {
+            await vs.DisposeAsync();
+        }
+    }
+}";
+
+            await Verify.VerifyAnalyzerAsync(test);
+        }
+
+        [Fact]
+        public async Task MethodChecksBoth_WithIsCheck()
+        {
+            var test = Preamble + @"
+class Test {
+    async Task CheckAndDispose(object o) {
+        if (o is BclAsyncDisposable) {
+            await ((BclAsyncDisposable)o).DisposeAsync();
+        }
+        else if (o is VsThreadingAsyncDisposable) {
+            await ((VsThreadingAsyncDisposable)o).DisposeAsync();
+        }
+    }
+}";
+
+            await Verify.VerifyAnalyzerAsync(test);
+        }
+
+        [Fact]
+        public async Task MethodChecksVsThreadingOnly_WithIsCheck()
+        {
+            var test = Preamble + @"
+class Test {
+    async Task CheckAndDispose(object o) {
+        if ([|o is VsThreadingAsyncDisposable|]) {
+            await ([|(VsThreadingAsyncDisposable)o|]).DisposeAsync();
```

Comment:
Do you notice the `[| ... |]` syntax in this test? That means a diagnostic is produced in those spans. So in fact this test passes because the analyzer finds this as a type check that is missing a check for the BCL type as well.